Fixed lexer grammar to parse numerics in scientific notation

phoebem opened the pull request:
Workflow condition parser throws exception when a number is written in scientific notation. This is a fix in the lex grammar to include it.

phoebem committed:
Fixed lexer grammar to parse numerics in scientific notation

phoebem committed:
resolve

phoebem committed:
Merge branch 'develop' into workflow_conditionparser_scientific_notation

lkiesow commented on modules/workflow-condition-parser/src/main/antlr4/org/opencastproject/workflow/conditionparser/antlr/WorkflowCondition.g4:
```diff
@@ -10,7 +10,7 @@ atom : '(' relationOperand ')' | NUMBER | STRING | BOOL ;
 OR: 'OR' ;
 NOT: 'NOT' ;
-NUMBER: '-'? [0-9]+ ('.' [0-9]+)? ;
+NUMBER: '-'? [0-9]+ ('.' [0-9]+)? ( [E|e] [+\-]? [0-9]+ )? ;
```
If you actually want to parse what Java allows you to parse, you might also want to allow a `+` sign in front of the first number:
```suggestion
NUMBER: [+\-]? [0-9]+ ('.' [0-9]+)? ( [E|e] [+\-]? [0-9]+ )? ;
```

lkiesow reviewed (approved):
Seems reasonable. Does make the lexer more complex and I don't really see a use case but I also don't really mind :D

Let me know if you want to add my suggestion or not.

phoebem committed:
Add '+' to number grammar

phoebem commented:
Thanks Lars,

 I like your suggestion. It makes it even more general and I updated my PR.

The reason we needed this was because I added bitrate to the analyze_tracks wf so that we can tune the encoding profiles according to width, height and bitrate.
When the bitrate was higher, I ended up with something like 1.0675E7 and the workflow choked.
This is a bit of a niche feature and I don’t think anyone would care about this upstream, so it is not in the PR, 
but the definition of number should include scientific notation, so I decided to make a pull against it.

Phoebe.


> On Oct 11, 2020, at 11:21 AM, Lars Kiesow <notifications@github.com> wrote:
> 
> WARNING: Harvard cannot validate this message was sent from an authorized system. Please be careful when opening attachments, clicking links, or following instructions. For more information, visit the HUIT IT Portal and search for SPF.
> 
> @lkiesow approved this pull request.
> 
> Seems reasonable. Does make the lexer more complex and I don't really see a use case but I also don't really mind :D
> 
> Let me know if you want to add my suggestion or not.
> 
> In modules/workflow-condition-parser/src/main/antlr4/org/opencastproject/workflow/conditionparser/antlr/WorkflowCondition.g4 <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_1936-23discussion-5Fr502929174&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=aGkliVkC8NCqER-kADXQg0IVXejhXMDdkLkcZzunOlQ&s=6gT4jYy1L7x0lddIu938Lz4opeVxrNvONhl_RFMfP-4&e=>:
> 
> > @@ -10,7 +10,7 @@ atom : '(' relationOperand ')' | NUMBER | STRING | BOOL ;
>  AND: 'AND' ;
>  OR: 'OR' ;
>  NOT: 'NOT' ;
> -NUMBER: '-'? [0-9]+ ('.' [0-9]+)? ;
> +NUMBER: '-'? [0-9]+ ('.' [0-9]+)? ( [E|e] [+\-]? [0-9]+ )? ;
> If you actually want to parse what Java allows you to parse, you might also want to allow a + sign in front of the first number:
> 
> ⬇️ Suggested change
> -NUMBER: '-'? [0-9]+ ('.' [0-9]+)? ( [E|e] [+\-]? [0-9]+ )? ;
> +NUMBER: [+\-]? [0-9]+ ('.' [0-9]+)? ( [E|e] [+\-]? [0-9]+ )? ;
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_1936-23pullrequestreview-2D506189652&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=aGkliVkC8NCqER-kADXQg0IVXejhXMDdkLkcZzunOlQ&s=pq_hiucVkqGcsT98pXG4DzVfxgOuo3hdWqP0NxJ5fk4&e=>, or unsubscribe <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABR7ENC6BCG3DYZADXL5VJDSKHEOVANCNFSM4SH5F7TQ&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=aGkliVkC8NCqER-kADXQg0IVXejhXMDdkLkcZzunOlQ&s=AaxgIMpauup6RsghaAIfiLnL0JQninz8MwO5R8Fv45Y&e=>.
>

lkiesow commented:
Ups, seems like my suggestion broke a test:
https://github.com/opencast/opencast/blob/70bacbfaa9c25f0f070543bff6eaca2472499cc6/modules/workflow-condition-parser/src/test/java/org/opencastproject/workflow/conditionparser/WorkflowConditionInterpreterTest.java#L86-L89

Though to be fair, something like `1-1` was probably already broken before

phoebem commented:
Hi Lars,

As you said, it was broken before my changes because it didn’t deal with negative numbers.
I spent a little time on it and conclude that to fix it properly is a bit more work than I have time right now.
I don’t know if anyone is using this module to parse math expressions, 
can we do one of the following:

1) Put in the documentation that you need to use spaces between numbers and numerical operators, add spaces to the tests.
2) Disallow negative numbers and remove the sign before number in the grammar.


Phoebe.
 

> On Oct 12, 2020, at 5:37 PM, Lars Kiesow <notifications@github.com> wrote:
> 
> WARNING: Harvard cannot validate this message was sent from an authorized system. Please be careful when opening attachments, clicking links, or following instructions. For more information, visit the HUIT IT Portal and search for SPF.
> 
> Ups, seems like my suggestion broke a test:
> https://github.com/opencast/opencast/blob/70bacbfaa9c25f0f070543bff6eaca2472499cc6/modules/workflow-condition-parser/src/test/java/org/opencastproject/workflow/conditionparser/WorkflowConditionInterpreterTest.java#L86-L89 <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_blob_70bacbfaa9c25f0f070543bff6eaca2472499cc6_modules_workflow-2Dcondition-2Dparser_src_test_java_org_opencastproject_workflow_conditionparser_WorkflowConditionInterpreterTest.java-23L86-2DL89&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=Fu6mNTrEwkY5Sm8eZa835TWn2BKUaILrIhoggvJEehU&s=pPdaAjvi6gi6r40WIwb9I9C-_DDhxnPzujTjKwvUPJ0&e=>
> Though to be fair, something like 1-1 was probably already broken before
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_1936-23issuecomment-2D707357720&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=Fu6mNTrEwkY5Sm8eZa835TWn2BKUaILrIhoggvJEehU&s=KYx0aGeaVnRy_2swzmWXB8FI0eLaSXPn67Mu3o5v7gw&e=>, or unsubscribe <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABR7ENBLNF22T2WTRN5JMV3SKNZJDANCNFSM4SH5F7TQ&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=Fu6mNTrEwkY5Sm8eZa835TWn2BKUaILrIhoggvJEehU&s=K6wj6sP297lcQ6vW8EhKHV8ACOkHfL_LhqlXmbvb0T4&e=>.
>

lkiesow commented:
Hi Phoebe, I think both ways will work. After all, I don't think that it's actually used. At least not much.
I don't really have a favorite.
I f I had to pick, I would probably choose the first option.
But feel free to decide on your own.

KatrinIhler commented:
@phoebem Sounds like this PR is just waiting on a minor fix, any idea when that is going to be ready? :)

KatrinIhler commented:
@phoebem Since this hasn't been worked on for 6 months, I'm gonna close this for now. Feel free to re-open once it's ready to be reviewed again! :)
